kubelet: eviction: add memcg threshold notifier to improve eviction responsiveness

sjenning opened the pull request:
This PR adds the ability for the eviction code to get immediate notification from the kernel when the available memory in the root cgroup falls below a user defined threshold, controlled by setting the `memory.available` siginal with the `--eviction-hard` flag.

This PR by itself, doesn't change anything as the frequency at which new stats can be obtained is currently controlled by the cadvisor housekeeping interval.  That being the case, the call to `synchronize()` by the notification loop will very likely get stale stats and not act any more quickly than it does now.

However, whenever cadvisor does get on-demand stat gathering ability, this will improve eviction responsiveness by getting async notification of the root cgroup memory state rather than relying on polling cadvisor.

@vishh @derekwaynecarr @kubernetes/rh-cluster-infra

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32577)

<!-- Reviewable:end -->

sjenning committed:
kubelet: eviction: add memcg threshold notifier to improve eviction responsiveness

derekwaynecarr commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -136,6 +139,52 @@ func (m *managerImpl) IsUnderDiskPressure() bool {
+
+	for _, threshold := range thresholds {
+		// only enable memcg threshold notification if a hard memory eviction limit is set
+		if threshold.Signal != SignalMemoryAvailable || threshold.GracePeriod != time.Duration(0) {
```
why special case grace period?

Are we not able to define two thresholds - one for the soft and one for that hard eviction config?  By getting notified on a soft transition, we can get more accurate grace period calculation.

sjenning replied:
this makes the code more complex in that we are tracking two notifiers now, but it can be done

derekwaynecarr replied:
is the complexity that we need to have multiple notifiers?  i think for a first pass, just having it for eviction-hard is fine, but in the long run, i would want both.  ideally, we could have the memcg notification that we crossed the threshold, and set a timer to wake up at the associated grace period and run sync at that point.

derekwaynecarr commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -136,6 +139,52 @@ func (m *managerImpl) IsUnderDiskPressure() bool {
+			continue
+		}
+
+		quantity := getThresholdQuantity(threshold.Value, observed.capacity, resource.DecimalSI).String()
```
the format is largely useless unless we toString this, but memory is always BinarySI

sjenning replied:
the issue is that, when writing to the ```event_control```, it understand 5M but not 5Mi.  am i doing it wrong?  i am doing a ```.String()``` on it.

derekwaynecarr commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -136,6 +139,52 @@ func (m *managerImpl) IsUnderDiskPressure() bool {
+
+		quantity := getThresholdQuantity(threshold.Value, observed.capacity, resource.DecimalSI).String()
+		// TODO: remove hardcoded root memory cgroup path
+		cgpath := "/sys/fs/cgroup/memory"
```
how often do people mount this to a different path?  i guess we could inject the mountpoints somehow in a follow-on.

sjenning replied:
i'm not sure. @vishh do you know of any environment that don't use this path for the memcg?

derekwaynecarr commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -136,6 +139,52 @@ func (m *managerImpl) IsUnderDiskPressure() bool {
+			return err
+		}
+
+		glog.Infof("registered memory notification for %s on %s at %s", cgpath, attribute, quantity)
```
i have been prefixing log messages with "eviction manager" for easy grep...

derekwaynecarr commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -136,6 +139,52 @@ func (m *managerImpl) IsUnderDiskPressure() bool {
+			for {
+				// when we get a threshold notification, run synchronize to take action
+				m.thresholdNotifier.WaitForNotification()
+				glog.Infof("memory threshold notification recieved")
```
i before e except after c

also prefix "eviction manager"

derekwaynecarr replied:
debated on merit of throwing an event here as well... thoughts?  we don't know if the direction is moving above or below the threshold, just that it was crossed in *some* direction?

sjenning replied:
correct, the only way to (kinda) know is to read usage_in_bytes after the notification and see if it is above or below the threshold.

with regard to throwing an event, who would be interested in such an event?

derekwaynecarr commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -136,6 +139,52 @@ func (m *managerImpl) IsUnderDiskPressure() bool {
+
+		glog.Infof("registered memory notification for %s on %s at %s", cgpath, attribute, quantity)
+		m.thresholdNotifier = thresholdNotifier
+		go m.thresholdNotifier.Start()
```
do we need to support passing a stopCh around?

derekwaynecarr replied:
not seeing where Stop() is called on teardown...

derekwaynecarr commented on pkg/kubelet/eviction/threshold_notifier.go:
```diff
@@ -0,0 +1,126 @@
+	"syscall"
+)
+
+type ThresholdNotifier interface {
```
Godoc please

vishh commented:
I'd like to review this, but will probably have to wait till next week. Apologies!

sjenning commented:
Updated with typo fixes, log prefix, godoc, pulled threshold setting out of constructor and into a SetThreshold() on the interface, added stopCh to Start() (even though i used wait.NeverStop, seemed to align with the convention).

Still outstanding: do we want to have a threshold for soft limit too?

derekwaynecarr reviewed (changes requested):
I think I want this factored in a way that lets us verify that no-matter what the `ThresholdNotifier`, the module actually responds to a notification being triggered.  right now, nothing is actually testable in unit testing / mock with the current structure.

sjenning commented:
@derekwaynecarr fixed nits and refactored createThresholdNotification() just to see what it looks like.  still thinking about how to refactor the notifier code for better testing.

ingvagabund commented on pkg/kubelet/eviction/threshold_notifier.go:
```diff
@@ -0,0 +1,149 @@
+	}
+	eventfd := int(efd)
+	if eventfd < 0 {
+		return nil, fmt.Errorf("eventfd call failed")
```
missing ``syscall.Close(watchfd)``  and ``syscall.Close(controlfd)`` ?

sjenning commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -32,7 +33,7 @@ import (
 )
 
-// managerImpl implements NodeStabilityManager
+// managerImpl implements Manager
```
yes, the type assertion is already below the struct def.  i was just fixing an inaccuracy in the in comment.

ingvagabund replied:
First line that got hidden, no argue about that :)

sjenning commented:
@ingvagabund fixed up. thanks for the review!

derekwaynecarr commented:
@vishh -- is this something you want to review ?

vishh commented:
Yes. Apologies for the delay. I will shepherd this PR through soon.

On Fri, Sep 30, 2016 at 8:49 AM, Derek Carr <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh> -- is this something you want to review
> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32577#issuecomment-250780225>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKJxKPGfHk_Gfzp9YAfRMLNGxOiqqks5qvS-MgaJpZM4J7x_r>
> .
>

sjenning commented:
@derekwaynecarr @vishh I reworked this PR today.  The notifier interface is simpler now with just a Start() function.  I put in notifiers for both soft and hard limits.  Overall, I think it made it simpler, more functional, and possibly better for testing.  I can add the testing once the approach is approved.

sjenning commented:
@k8s-bot gci gce e2e test this

sjenning commented:
@k8s-bot node e2e test this issue #31408

derekwaynecarr commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -143,6 +146,34 @@ func (m *managerImpl) IsUnderInodePressure() bool {
+		if !found {
+			continue
+		}
+		// TODO: remove hardcoded root memory cgroup path
```
use: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/helpers_linux.go#L26 to find the mount point.

derekwaynecarr commented:
I like the new factoring, one nit and one request to use a common helper function.

derekwaynecarr commented on pkg/kubelet/eviction/threshold_notifier.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// ThresholdNotifierHandlerFunc is called when a threshold is crossed
+type ThresholdNotifierHandlerFunc func()
```
move this into pkg/eviction/types.go and make it `thresholdNotifierHandlerFunc`

sjenning commented:
@k8s-bot gci gke e2e test this

sjenning commented:
@k8s-bot node e2e test this

vishh commented on pkg/kubelet/eviction/threshold_notifier.go:
```diff
@@ -0,0 +1,114 @@
+package eviction
+
+/*
+#include <sys/eventfd.h>
```
cc @luxas I hope this does not impact static linking of kubelet

luxas replied:
I think it shouldn't, but I'm not 100% sure; the way we're going to compile kubelet is with glibc, but the glibc library we're using should be linked into the binary

Will check it up

vishh commented on pkg/kubelet/eviction/threshold_notifier.go:
```diff
@@ -0,0 +1,114 @@
+	}
+	efd, err := C.eventfd(0, C.EFD_CLOEXEC)
+	if err != nil {
+		syscall.Close(watchfd)
```
nit: Why not place this section in a `defer` block that is gated on `err != nil` ?

vishh commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -172,8 +210,27 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
+		m.notifiersInitialized = true
+		// start soft memory notification
+		startMemoryThresholdNotifier(m.config.Thresholds, observations, false, func() {
+			glog.Infof("soft memory eviction threshold crossed")
```
nit: Can we also include the %tage or absolute value here? It will be helpful while debugging.

vishh commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -172,8 +210,27 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
+	if !m.notifiersInitialized {
+		m.notifiersInitialized = true
+		// start soft memory notification
+		startMemoryThresholdNotifier(m.config.Thresholds, observations, false, func() {
```
You don't seem to be capturing the return error here.

vishh commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -172,8 +210,27 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
+			m.synchronize(diskInfoProvider, podFunc)
+		})
+		if err != nil {
+			glog.Warningf("eviction manager: failed to create hard memory threshold notifier: %v", err)
```
Why not retry again if it fails?

vishh commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -172,8 +210,27 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
+		}
+		// start hard memory notification
+		startMemoryThresholdNotifier(m.config.Thresholds, observations, true, func() {
+			glog.Infof("hard memory eviction threshold crossed")
```
ditto here - include threshold value in the log message

vishh commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -145,6 +150,39 @@ func (m *managerImpl) IsUnderInodePressure() bool {
+func startMemoryThresholdNotifier(thresholds []Threshold, observations signalObservations, hard bool, handler thresholdNotifierHandlerFunc) error {
+	for _, threshold := range thresholds {
+		if threshold.Signal != SignalMemoryAvailable ||
+			(hard && threshold.GracePeriod != time.Duration(0)) ||
```
AFAIK, we don't support eviction grace period for `hard` thresholds. We only support `termination grace period`. cc @derekwaynecarr

sjenning replied:
@vishh my understanding is the if `threshold.GracePeriod == 0`, the threshold is a hard limit.

derekwaynecarr replied:
@sjenning your understanding is correct.

vishh replied:
Can you add a few comments describing the intent of this logic here? It will help us in the future.

vishh commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -145,6 +150,39 @@ func (m *managerImpl) IsUnderInodePressure() bool {
+		if err != nil {
+			return err
+		}
+		cgpath, found := cgroups.MountPoints["memory"]
```
note that in the future, we plan to support evictions both on root and on `--cgroups-root`. May be its worth adding a TODO here or even better file an issue for it.

sjenning commented:
@vishh thanks for the review! I have updated the PR.

sjenning commented:
@k8s-bot node e2e test this

sjenning commented:
@k8s-bot node e2e test this

derekwaynecarr commented:
@k8s-bot node e2e test this

sjenning commented:
@k8s-bot gci gke e2e test this

vishh commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -145,6 +150,41 @@ func (m *managerImpl) IsUnderInodePressure() bool {
+			return err
+		}
+		// TODO add support for eviction from --cgroup-root
+		cgpath, found := cgroups.MountPoints["memory"]
```
nit: I'd prefer moving all cgroups logic to under `/pkg/kubelet/cm` to have better abstractions. I'm OK to perform that refactor in a separate PR though given that v1.5 FF is this week.

vishh reviewed (changes requested):
Just one more `nit`, and then this is good to go. I will add LGTM with a `do-not-merge` label. Feel free to remove the label once you fix the nit.

derekwaynecarr commented:
@sjenning -- can you rebase and fix the nit?

sjenning commented:
@derekwaynecarr rebased. i think @vishh said he was ok will pulling the cgroups stuff out in a different PR.  want to change as little as possible at this point as to not introduce new points of contention.

sjenning commented:
@k8s-bot node e2e test this

dims commented:
@k8s-bot node e2e test this

vishh commented:
@k8s-bot node e2e test this

vishh commented:
@k8s-bot node e2e test this

vishh commented:
@Random-Liu This PR passes node e2e locally and remotely. Note sure why it fails Node e2e entirely.

Random-Liu commented:
https://gist.github.com/Random-Liu/d1d2d9d245b130576867425cc1932da9

Random-Liu commented:
It seems that the PR makes ubuntu trusty node into bad state. After the test the node is not accessible any more, and [this line](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/remote/remote.go#L314) stuck forever because of lack of ssh timeout until the jenkins job is killed forcibly.

The nodes are still there now. I tried to ssh, and the ssh pending forever, which proved my theory.

We should:
1) Figure out why this PR breaks ubuntu trusty.
2) Improve node e2e to add connection timeout for all ssh command.

@vishh @derekwaynecarr @sjenning

sjenning commented:
@vishh working on this now

sjenning commented:
my only thoughts at the moment are 1) cgroup incompatibility pre-systemd or 2) some eventfd incompatibility

sjenning commented:
@vishh @Random-Liu @derekwaynecarr 

The ubuntu trusty kernel is having a big problem.  From the serial output on the GCE console, since ssh is basically being DoSed:

```
[  455.437838] BUG: Bad page state in process khugepaged  pfn:b5972
[  455.443955] page:ffffea0002d65c80 count:1 mapcount:0 mapping:ffff8800bb314438 index:0x0
[  455.452066] flags: 0x1ffff000008003c(referenced|uptodate|dirty|lru|swapbacked)
[  455.460077] page dumped because: page still charged to cgroup
[  455.465945] bad because of flags:
[  455.469363] flags: 0x20(lru)
[  455.472579] page->mem_cgroup:ffff88012b485000
```

They are never ending.  GCE console shows the CPU is pegged.  There seems to be a bug in the memcg implementation for the trusty kernel that my PR is hitting.

sjenning commented:
I did this a few more times last night and the test runs completed, mind you, with tons of failures.  The number failures varied.  I had large numbers of failures even with running against master so, it appears that not all e2e node test invocations are equal.

I'm doing this
```
make test-e2e-node REMOTE=true IMAGE_PROJECT=kubernetes-node-e2e-images IMAGES=e2e-node-ubuntu-trusty-docker10-v1-image
```

@vishh Should I be doing something else?

vishh commented:
If we can understand the reason for the failure, it might be possible to
disable the test just on Ubuntu.
Users might hit the same bug and we need to warn them and so triaging the
failures is imperative.

On Wed, Nov 16, 2016 at 9:18 AM, Seth Jennings <notifications@github.com>
wrote:

> I did this a few more times last night and the test runs completed, mind
> you, with tons of failures. The number failures varied. I had large numbers
> of failures even with running against master so, it appears that not all
> e2e node test invocations are equal.
>
> I'm doing this
>
> make test-e2e-node REMOTE=true IMAGE_PROJECT=kubernetes-node-e2e-images IMAGES=e2e-node-ubuntu-trusty-docker10-v1-image
>
> @vishh <https://github.com/vishh> Should I be doing something else?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32577#issuecomment-261010049>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKJwkFiOdNLCEF96upzG2xlmluSeLks5q-zsCgaJpZM4J7x_r>
> .
>

sjenning commented:
@vishh @derekwaynecarr 

I created a pristine ubuntu trusty cloud instance (Daily Build [20161111]) with the same amount of vCPU and RAM as an `n1-standard-1`, which is size the CI runs the e2e-node instances.  I ran the e2e node tests with `make test-e2e-node SKIP=Slow TEST_ARGS='--feature-gates=DynamicKubeletConfig=true' RUN_UNTIL_FAILURE=true`.  It completed 5 completely successful passes before flaking on

```
[Fail] [k8s.io] Summary API when querying /stats/summary [It] should report resource usage through the stats api 
/root/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/summary_test.go:208
```

I was not able to reproduce the node hang.  It is possible that a kernel update could fix the bug that causes the page table corruption.

I think the way forward here is to update the ubuntu trusty e2e-node image.

vishh commented:
@mtaufen @Random-Liu can one you help @sjenning update our ubuntu trusty node e2e test image?

dims commented:
@vishh @dchen1107 : would you classify this as a release-blocker?

derekwaynecarr commented:
@dims -- it corresponds to https://github.com/kubernetes/features/issues/125

dims commented:
@derekwaynecarr Ack. let me mark it as release-blocker

dchen1107 commented:
@Random-Liu Can you help with ubuntu image here?

dchen1107 commented:
I filed a separate issue #37251 to automate the image build process. And @Random-Liu is working on building the new image for this. Sorry for the delay here.

Random-Liu commented:
@dchen1107 Sure. I'll do this.

derekwaynecarr commented:
@k8s-bot test this ISSUE #ignore 

(hope this picks up new images)

derekwaynecarr commented:
@k8s-bot bazel test this

derekwaynecarr commented:
@mikedanese -- are you able to help instruct what is needed to make bazel build happy?

derekwaynecarr commented:
I am guessing hack/update-bazel does not work well with cgo stuff, and we would have to hand edit this file.  @sjenning -- not sure if you are around tomorrow / tonight to investigate more.

yujuhong commented:
@k8s-bot gce etcd3 e2e test this

mikedanese commented:
You found a bug in gazel (or maybe an unimplemented feature)! https://github.com/mikedanese/gazel/pull/3is under review and will fix the bazel build failure. Again the bazel presubmit is not currently blocking. As you can see the Submit Queue check is green even though Bazel Build is failing.

derekwaynecarr commented:
@mikedanese - thanks!

rmmh commented:
@dchen1107 @mikedanese @sjenning this broke the cross-build, but I'm don't understand why. Please revert or fix.

mikedanese commented:
Its likely due to the CGO.

derekwaynecarr commented:
@rmmh -- I am putting together a fix for cross-build now.

saad-ali commented:
Removing cherry-pick candidate label this is already in 1.5

saad-ali commented:
@sjenning @derekwaynecarr @vishh  This PR is currently on trial for causing issue https://github.com/kubernetes/kubernetes/issues/37853. @dchen1107, @mtaufen, and @thockin are working to verify. If convicted, this PR will be reverted from master and the 1.5 release branch.

Could you help me understand what the impact of reverting this PR and shipping 1.5.0 without it will be?

derekwaynecarr commented:
Can we gate its enablement behind an additional flag?  It's unclear if this impacted all kernels and it's not clear to me if it only had an impact when you experience memory pressure.  Will speak to Seth more in morning.  The impact is Kubelet is less responsive to memory pressure and node is more likely to OOM assuming kernel itself was stable with the notification.

derekwaynecarr commented:
For example, a --kernel-memcg-notify flag would be a simple way to gate the behavior on impacted distros.

vishh commented:
This PR will help avoid production issues that occur often due to Node
running into Memory Pressure. This improves the reliability of the node. So
before reverting, it would help if we can further understand the impact of
this PR and identify any potential fixes for the base image.

On Wed, Dec 7, 2016 at 7:56 AM, Derek Carr <notifications@github.com> wrote:

> Can we gate its enablement behind an additional flag? It's unclear if this
> impacted all kernels and it's not clear to me if it only had an impact when
> you experience memory pressure. Will speak to Seth more in morning. The
> impact is Kubelet is less responsive to memory pressure and node is more
> likely to OOM assuming kernel itself was stable with the notification.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32577#issuecomment-265337146>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKLl2IXpcSeQTe6Zujnqmf4U_oUYrks5rFhlngaJpZM4J7x_r>
> .
>

saad-ali commented:
> This PR will help avoid production issues that occur often due to Node running into Memory Pressure. This improves the reliability of the node. So before reverting, it would help if we can further understand the impact of this PR and identify any potential fixes for the base image.

Could you please help enumerate those.

vishh commented:
Kubernetes nodes are typically memory overcommitted by default unless Node Allocatable is configured. This results in nodes invoking the kernel OOM killer which ends up killing system services like docker daemon, kubelet, kube-proxy at times, in addition to arbitrary user processes. We introduced a user space eviction logic in kubelet to avoid invoking kernel OOM killer as much as possible by evicting user pods to free up memory. 
Prior to this PR, that user space eviction logic ran in a housekeeping loop and missed memory spikes at times, and failed to perform its intended functionality effectively. 
With this PR, that user space memory eviction logic will be triggered by the kernel whenever there is memory pressure and evict pods immediately. 
@saad-ali I can enumerate a list of customers hitting this issue offline.

derekwaynecarr commented:
@sjenning @vishh @saad-ali - i have started to put a patch to gate its enablement behind an additional flag for kernels where the notification could prove problematic.

dchen1107 commented:
I am ok to have a flag to gate the feature.

vishh commented:
Thanks @derekwaynecarr

derekwaynecarr commented:
opened a WIP pr here for a flag:  https://github.com/kubernetes/kubernetes/pull/38258

hope to finish this evening.
